docs: add `artifactType` to signature spec (#325)
notaryproject/specifications: specs/signature-specification.md

## OCI Signatures

This section describes how a Notary Project signature is stored in an OCI Distribution conformant registry.
OCI image manifest is used to store signatures in the registry, see [OCI image spec v1.1.1][oci-image-manifest] for details. 

The signature manifest uses either the `artifactType` property or `config.mediaType` to indicate it's a Notary Project signature, a subject referencing the manifest of the artifact being signed, a layer referencing the signature, and a collection of annotations.

![Signature storage inside registry](../media/new-oci-signature-specification.svg)

> - For registries compliant with OCI Distribution Spec v1.1, the Notary Project signature manifest by default uses the required `artifactType` property to indicate it's a Notary Project signature when `config.mediaType` is empty. The signing agent is recommended to produce the signature with `artifactType` property by default. The sample signature manifest is as follows:

Reviewer: We should mention the value `application/vnd.cncf.notary.signature` here.

## OCI Signatures

This section describes how a Notary Project signature is stored in an OCI Distribution conformant registry.
OCI image manifest is used to store signatures in the registry, see [OCI image spec v1.1.0][oci-image-manifest] for details. Notary Project signature is compliant with OCI image spec v1.1.0.

> Starting from `notation` v2.0.0, the signature manifest uses the `artifactType` property that specifies it's a Notary Project signature, a subject referencing the manifest of the artifact being signed, a layer referencing the signature, and a collection of annotations.

Reviewer: Again, this is the Notary Project spec for all implementations. Details regarding `Notation v2.0.0` should be put into Notation CLI's docs/spec.

Besides the [image manifest property requirements][image-manifest-property-descriptions], the properties have the following additional restrictions:

- **`mediaType`** (*string*): This REQUIRED property MUST be `application/vnd.oci.image.manifest.v1+json`.

> - **`artifactType`** (*string*): This OPTIONAL property MUST be `application/vnd.cncf.notary.signature`.

Reviewer: @FeynmanZhou your change doesn't align with my original proposed solution. I left a detailed comment in the original discussion. PTAL.

- **`config`** (*descriptor*): This property is REQUIRED to be compatible with [OCI image specification][oci-image-manifest]. The Notary Project signature specification doesn't require any configuration for a signature, and the configuration content is not consumed by implementations of the Notary Project signature specification.

>   - **`mediaType`** (*string*): This REQUIRED property MUST be `application/vnd.oci.empty.v1+json`.

Reviewer: > Another possibility: signing migrate to `artifactType`, but verification supports both `Config.MediaType` and `artifactType` for backward compatibility.

That's what I mean. For `signing`, all new signatures should use the `artifactType` field.

>   - **`mediaType`** (*string*): This REQUIRED property MUST be `application/vnd.oci.empty.v1+json`.

Reviewer: @FeynmanZhou Your most recent update doesn't seem to align with my original proposal. Below is incorrect per OCI image spec:
```
**`mediaType`** (*string*): This REQUIRED property MUST be `application/vnd.oci.empty.v1+json`
```
Please read through the proposed solution again.

More specifically, the `artifactType` field is REQUIRED when `config.MediaType` is empty. When artifactType does not exist OR its value is not application/vnd.cncf.notary.signature, fall back to use Config.MediaType, which must be set to application/vnd.cncf.notary.signature.

Besides the [image manifest property requirements][image-manifest-property-descriptions], the properties have the following additional restrictions:

- **`mediaType`** (*string*): This REQUIRED property MUST be `application/vnd.oci.image.manifest.v1+json`.

> - **`artifactType`** (*string*): This REQUIRED property MUST be `application/vnd.cncf.notary.signature`.

Reviewer: `artifactType` shouldn't be REQUIRED, otherwise, it's a breaking change to Notary Project spec v1.
It should be OPTIONAL, when it's missing, `config.MediaType` is REQUIRED and the value MUST be `application/vnd.cncf.notary.signature`.

- **`config`** (*descriptor*): This property is REQUIRED to be compatible with [OCI image specification][oci-image-manifest]. The Notary Project signature specification doesn't require any configuration for a signature, and the configuration content is not consumed by implementations of the Notary Project signature specification.

>   - **`mediaType`** (*string*): This REQUIRED property MUST be `application/vnd.oci.empty.v1+json`.

Reviewer: This is a breaking change to Notary Project spec v1.

>   - **`mediaType`** (*string*): This property is REQUIRED. Its value MUST be `application/vnd.cncf.notary.signature` when `artifactType` does not exist. If `artifactType` is set to `application/vnd.cncf.notary.signature`, uses `application/vnd.oci.empty.v1+json` as its value.

Reviewer: ```suggestion
  - **`mediaType`** (*string*): This property is REQUIRED. Its value MUST be `application/vnd.cncf.notary.signature` when `artifactType` does not exist. When `artifactType` exists, `mediaType` MUST be `application/vnd.oci.empty.v1+json`.
```

For OCI v1.0 compliant registries, the `subject` and `artifactType` fileds in the signature manifest are ignored by the registry. Notation v2.x SHOULD fall back to use OCI referrers tag schema to push and store the signature in the registry.

#### OCI Signature Verification Forward and Backward Compatibility  

The Notary Project signature uses the `artifactType` property to indicate its type starting from `notation` v2.0.0, whereas older versions rely on `config.mediatype`.  

To minimize the impact of breaking changes in the signature manifest, `notation` v2.x SHOULD be able to list, inspect, and verify the signature produced by `notation` v1.x. It ensures backward compatibility with older OCI signatures.

> Forward compatibility SHOULD also be considered in `notation` v1.x since it reduces friction in updrades and won't make older workflows suddenly fail when verifying a newer signature. `notation` v1.x SHOULD be able to list, inspect, and verify the signature produced by `notation` v2.0. 

Reviewer: Same as above, this is the Notary Project spec repo. Discussions on Notation should happen in the Notation repo.

  Keys using the `io.cncf.notary` namespace are reserved for use in the Notary Project signature specification and MUST NOT be used by other specifications.
  - **`io.cncf.notary.x509chain.thumbprint#S256`**: A REQUIRED annotation whose value contains the list of SHA-256 fingerprints of signing certificate and certificate chain (including root) used for signature generation. The list of fingerprints is present as a JSON array string, corresponding to ordered certificates in [*Certificate Chain* unsigned attribute](#unsigned-attributes) in the signature envelope. The annotation name contains the hash algorithm as a suffix (`#S256`) and can be extended to support other hashing algorithms in future.

> - For registries compliant with OCI Distribution Spec v1.0, the `subject` and `artifactType` fileds in the signature manifest are ignored by the registry. The signing agent SHOULD fall back to use OCI referrers tag schema to produce a Notary Project signature using a legacy format which does not contain `artifactType` property. 

Reviewer: ```suggestion
- For registries compliant with OCI Distribution Spec v1.0, the `subject` and `artifactType` fields in the signature manifest are ignored by the registry. The signing agent SHOULD fall back to use OCI referrers tag schema to produce a Notary Project signature using a legacy format which does not contain `artifactType` property. 
```

```jsonc
{
    "schemaVersion": 2,
    "mediaType": "application/vnd.oci.image.manifest.v1+json",
    "config": {
        "mediaType": "application/vnd.cncf.notary.signature",
        "size": 2,
        "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a"
    },
    "layers": [
        {
            "mediaType": "application/jose+json",
            "digest": "sha256:9834876dcfb05cb167a5c24953eba58c4ac89b1adf57f28f2f9d09af107ee8f0",
            "size": 32654
        }
    ],
    "subject": {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "digest": "sha256:73c803930ea3ba1e54bc25c2bdc53edd0284c62ed651fe7b00369da519a3c333",
        "size": 16724
    },
    "annotations": {
        "io.cncf.notary.x509chain.thumbprint#S256": 
        "[\"B7A69A70992AE4F9FF103EBE04A2C3BA6C777E439253CE36562E6E98375068C3\",\"932EB6F5598435D4EF23F97B0B5ACB515FAE2B8D8FAC046AB813DDC419DD5E89\"]"
    }
}
```

#### OCI Signature Verification Forward and Backward Compatibility  

When signing an artifact, the Notary Project signature uses the `artifactType` property to indicate its type, whereas older versions rely on `config.mediatype`.  

> To minimize the impact of breaking changes in the signature manifest, verifier SHOULD be able to list, inspect, and verify either the new or legacy format of Notary Project signature. It ensures forward and backward compatibility with OCI signatures. It reduces friction in upgrades and won't make older workflows suddenly fail when verifying a newer signature. 

Reviewer: Just an open discussion:
I think this design is still not backward compatible though. If a new signature manifest is using the `artifactType` way, then an old Notary Project implementation cannot take it as a notary signature? Because any old implementation of this spec is expecting the `config.MediaType`.
But I guess the above is not really an issue? (the signer has the option to create the signature with the old way.)

Reviewer: Based on our offline discussion, verifier will need to support verifying both new and legacy format signatures. 
Please feel free to resolve this comment.
